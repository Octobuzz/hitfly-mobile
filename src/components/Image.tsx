import React from 'react'
import FastImage, { FastImageProperties } from 'react-native-fast-image'
// @ts-ignore https://github.com/react-native-community/react-native-svg/issues/1089
import { SvgXml } from 'react-native-svg'

const fetchText = async (uri: string): Promise<string> => {
  try {
    const response = await fetch(uri)
    return await response.text()
  } catch {
    return ''
  }
}

const removeStylesFromSvg = (svg: string): string => {
  const tmp = svg.replace(/<style>.*<\/style>/gi, '')
  console.tron.log(tmp)
  return tmp
}

const Image: React.FC<FastImageProperties> = ({
  source,
  style = { width: '100%', height: '100%' },
  ...rest
}) => {
  const [xml, setXml] = React.useState()
  React.useEffect(() => {
    if (
      typeof source !== 'number' &&
      source.uri &&
      source.uri.endsWith('.svg')
    ) {
      fetchText(source.uri)
        .then(removeStylesFromSvg)
        .then(setXml)
    }
  }, [source])

  return xml ? (
    <SvgXml style={style} xml={xml} />
  ) : (
    <FastImage source={source} style={style} {...rest} />
  )
}
export default Image
