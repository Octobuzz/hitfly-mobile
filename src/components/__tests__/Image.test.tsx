import React from 'react'
import Image from '../Image'
import renderer from 'react-test-renderer'

test('renders correctly with number source', () => {
  const tree = renderer.create(<Image source={1} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders correctly with not svg source', () => {
  const tree = renderer.create(<Image source={{ uri: 'not svg' }} />).toJSON()
  expect(tree).toMatchSnapshot()
})
