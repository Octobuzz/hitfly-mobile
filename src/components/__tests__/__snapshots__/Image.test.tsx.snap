// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with not svg source 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={
      Object {
        "uri": "not svg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`renders correctly with number source 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={null}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`renders correctly with svg source 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={
      Object {
        "uri": ".svg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
